plugin/core/sessions.py

```diff
@@ -2480,10 +2480,11 @@ def send_request_task_2(self, request: Request) -> tuple[Promise, int]:
         request_id = self.send_request_async(request, resolver, lambda x: resolver(Error.from_lsp(x)))
         return (promise, request_id)
 
-    def cancel_request(self, request_id: int, ignore_response: bool = True) -> None:
+    def cancel_request_async(self, request_id: int, *, ignore_response: bool = True) -> None:
         self.send_notification(Notification("$/cancelRequest", {"id": request_id}))
-        if ignore_response and request_id in self._response_handlers:
-            request, _, _ = self._response_handlers[request_id]
+        request, _, _ = self._response_handlers[request_id]
```

Comment:
This should better be a `self._response_handlers.get(request_id)`

The entry for the `request_id` gets pop'ed when the response arrives, so I think this line would be prone to a race condition. That is also the reason why I would keep it as part of the if-condition below.